http3/http_stream.go

```diff
@@ -68,19 +77,45 @@ func (s *stream) Read(b []byte) (int, error) {
 		conn: s.conn,
 	}
 	if s.bytesRemainingInFrame == 0 {
+		var seenTrailerFrame bool
 	parseLoop:
 		for {
 			frame, err := fp.ParseNext()
 			if err != nil {
 				return 0, err
 			}
 			switch f := frame.(type) {
-			case *headersFrame:
-				// skip HEADERS frames
-				continue
 			case *dataFrame:
+				if seenTrailerFrame {
+					return 0, errors.New("DATA frame received after trailers")
+				}
 				s.bytesRemainingInFrame = f.Length
 				break parseLoop
+			case *headersFrame:
+				if s.conn.perspective == protocol.PerspectiveServer {
+					continue
+				}
+				if f.Length > s.maxHeaderBytes {
+					return 0, fmt.Errorf("HEADERS frame too large: %d bytes (max: %d)", f.Length, s.maxHeaderBytes)
+				}
+				if seenTrailerFrame {
+					return 0, errors.New("HEADERS frame received after trailers")
+				}
+				p := make([]byte, f.Length)
+				s.Stream.Read(p)
+				trailers, err := s.decoder.DecodeFull(p)
+				if err != nil {
+					// Ignoring invalid frame
```

Comment:
Why? This seems wrong.

Reply:
Yeah, I agree. This was a holdover from the first PR that I didn't quite understand the reasoning behind. I've adjusted it to return the error instead of ignoring it and adjusted the associated test